### PR TITLE
send regular GPU utilization report with CronJob

### PR DESCRIPTION
**Pull request by suiguoxin**
- create a k8s CronJob which :
    - query `Prometheus` regularly for GPU utilization data
    - send cluster-usage alerts to `alert-manager`
    - the alerts will be handled by `alert-handler`

This feature was previously implemented by @Binyang2014 [here](https://github.com/microsoft/pai/blob/binyli/alert/src/alert-manager/deploy/alert-templates-notice/notice.tmpl), here we just adapt the code into `alert-manager`.

The alert mail:
image

### CODE REVIEW

File: src/alert-manager/deploy/alert-manager-cronjob.yaml.template [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+# DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+
+{% if cluster_cfg["alert-manager"]["use-pylon"] %}
+{% set external_url = cluster_cfg["pylon"]["uri"] %}
+{% else %}
+{% set external_url = cluster_cfg["alert-manager"]["url"] %}
+{% endif %}
```
Comment (Binyang2014) [inline src/alert-manager/deploy/alert-manager-cronjob.yaml.template:23]: Why need these lines?

File: src/alert-manager/src/cluster-utilization/send_alert.py [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+    job_infos = {}
+    job_usage = []
+    job_list = []
+    # get all jobs
+    headers = {'Authorization': "Bearer {}".format(TOKEN)}
+    resp = requests.get(rest_url, headers=headers)
+    if resp.ok:
+        job_list = resp.json()
+    else:
+        logging.warning("Failed to get job list")
```
Comment (Binyang2014) [inline src/alert-manager/src/cluster-utilization/send_alert.py:46]: Throw exception here?

File: src/alert-manager/src/cluster-utilization/send_alert.py [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+            logging.warning("request failed %s", resp.text)
+            continue
+        resp_json = resp.json()
+        if not resp_json["jobStatus"]["appLaunchedTime"]:
+            logging.warning("job not start, ignore it")
+            continue
+        job_info["start_time"] = datetime.fromtimestamp(
+            int(resp_json["jobStatus"]["appLaunchedTime"]) / 1000,
+            timezone.utc)
+        if not resp_json["jobStatus"]["appLaunchedTime"] or not resp_json[
```
Comment (Binyang2014) [inline src/alert-manager/src/cluster-utilization/send_alert.py:63]: Do not need to check `resp_json["jobStatus"]["appLaunchedTime"]` here?

**Review by Binyang2014 — COMMENTED** [review]
Can admin disable this report by comment the config?

File: src/alert-manager/src/cluster-utilization/send_alert.py (hunk 3 above)
Comment (suiguoxin) [reply src/alert-manager/src/cluster-utilization/send_alert.py]: > Do not need to check `resp_json["jobStatus"]["appLaunchedTime"]` here?

Removed, thx

File: src/alert-manager/src/cluster-utilization/send_alert.py (hunk 2 above)
Comment (suiguoxin) [reply src/alert-manager/src/cluster-utilization/send_alert.py]: > Throw exception here?

Fixed, thx

File: src/alert-manager/deploy/alert-manager-cronjob.yaml.template (hunk 1 above)
Comment (suiguoxin) [reply src/alert-manager/deploy/alert-manager-cronjob.yaml.template]: > Why need these lines?

Removed, thanks

**Comment by suiguoxin** [comment]
> Can admin disable this report by comment the config?

The code is refactored. After this change, the report will only be enabled when the `schedule` field is set.

**Review by Binyang2014 — APPROVED** [review]

**Pull request merged**